From c1d4fc64b5dcf8f414f424bea7fe3c1655c32634 Mon Sep 17 00:00:00 2001
From: Mike Snitzer <snitzer@redhat.com>
Date: Fri, 13 Jun 2014 14:36:40 -0400
Subject: [block] blk-mq: don't allow queue entering for a dying queue

Message-id: <1402670238-13916-113-git-send-email-snitzer@redhat.com>
Patchwork-id: 83855
O-Subject: [RHEL7.1 PATCH 112/150] blk-mq: don't allow queue entering for a dying queue
Bugzilla: 1105204
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Jeff Moyer <jmoyer@redhat.com>

BZ: 1105204

Upstream commit 3b632cf0eaa2e89a12c18f043e6e7c5bcc003645
Author: Keith Busch <keith.busch@intel.com>
Date:   Fri Jun 6 10:22:07 2014 -0600

    blk-mq: don't allow queue entering for a dying queue

    If the queue is going away, don't let new allocs or queueing
    happen on it. Go through the normal wait process, and exit with
    ENODEV in that case.

    Signed-off-by: Keith Busch <keith.busch@intel.com>
    Signed-off-by: Jens Axboe <axboe@fb.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/block/blk-mq.c b/block/blk-mq.c
index a6ee74e..75fc33f 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -82,8 +82,10 @@ static int blk_mq_queue_enter(struct request_queue *q)
 
  __percpu_counter_add(&q->mq_usage_counter, 1, 1000000);
  smp_wmb();
- /* we have problems to freeze the queue if it's initializing */
- if (!blk_queue_bypass(q) || !blk_queue_init_done(q))
+
+ /* we have problems freezing the queue if it's initializing */
+ if (!blk_queue_dying(q) &&
+     (!blk_queue_bypass(q) || !blk_queue_init_done(q)))
   return 0;
 
  __percpu_counter_add(&q->mq_usage_counter, -1, 1000000);
-- 
1.7.1